src/common/ad_mem.c

```diff
@@ -3,5 +3,2003 @@
+		}
+	} else {
+		D_ASSERT(0); /* XXX remove this while integrating with other components. */
+	}
+
+	/* NB: ad_blob_df (superblock) is stored right after header of arena[0],
+	 * so it does not require any special code to handle checkpoint of superblock.
+	 */
+	blob->bb_df = (struct ad_blob_df *)&blob->bb_pages[0].pa_rpg[ARENA_HDR_SIZE];
+
+	/* bitmap for reserving arena */
+	D_ALLOC(blob->bb_bmap_rsv, blob_bmap_size(blob));
+	if (!blob->bb_bmap_rsv)
+		goto failed;
+
+	for (i = 0; i < ARENA_LRU_SIZE; i++) {
+		struct ad_arena *arena;
+
+		arena = alloc_arena(NULL, true);
+		if (!arena)
+			goto failed;
+
+		blob->bb_ars_lru_size++;
+		d_list_add(&arena->ar_link, &blob->bb_ars_lru);
+	}
+
+	for (i = 0; i < GROUP_LRU_SIZE; i++) {
+		struct ad_group *group;
+
+		group = alloc_group(NULL, true);
+		if (!group)
+			goto failed;
+
+		blob->bb_gps_lru_size++;
+		d_list_add(&group->gp_link, &blob->bb_gps_lru);
+	}
+	return 0;
+failed:
+	blob_fini(blob);
+	return -DER_NOMEM;
+}
+
+static int
+blob_fini(struct ad_blob *blob)
+{
+	struct ad_arena *arena;
+	struct ad_group *group;
+
+	D_DEBUG(DB_TRACE, "Finalizing blob\n");
+	while ((group = d_list_pop_entry(&blob->bb_gps_rsv, struct ad_group, gp_link))) {
+		D_ASSERT(group->gp_unpub);
+		group_decref(group);
+	}
+	while ((arena = d_list_pop_entry(&blob->bb_ars_rsv, struct ad_arena, ar_link))) {
+		D_ASSERT(arena->ar_unpub);
+		arena_decref(arena);
+	}
+
+	while ((group = d_list_pop_entry(&blob->bb_gps_lru, struct ad_group, gp_link)))
+		free_group(group, true);
+	while ((arena = d_list_pop_entry(&blob->bb_ars_lru, struct ad_arena, ar_link)))
+		free_arena(arena, true);
+
+	if (blob->bb_bmap_rsv)
+		D_FREE(blob->bb_bmap_rsv);
+	if (blob->bb_pages)
+		D_FREE(blob->bb_pages);
+	if (blob->bb_mmap)
+		D_FREE(blob->bb_mmap);
+	return 0;
+}
+
+static int
+blob_load(struct ad_blob *blob)
+{
+	struct ad_blob_df *bd = blob->bb_df;
+	int		   i;
+	int		   rc = 0;
+
+	for (i = 0; i < blob->bb_pgs_nr; i++) {
+		struct ad_page		*page;
+		struct umem_store_iod	 iod;
+		d_sg_list_t		 sgl;
+		d_iov_t			 iov;
+
+		page = &blob->bb_pages[i];
+		ad_iod_set(&iod, blob_addr(blob) + ARENA_SIZE * i, ARENA_SIZE);
+		ad_sgl_set(&sgl, &iov, page->pa_rpg, ARENA_SIZE);
+
+		/* XXX: submit multiple pages, otherwise it's too slow */
+		rc = blob->bb_store.stor_ops->so_read(&blob->bb_store, &iod, &sgl);
+		if (rc) {
+			D_ERROR("Failed to load storage contents: %d\n", rc);
+			goto out;
+		}
+	}
+
+	/* NB: @bd points to the first page, its content has been brought in by the above read.*/
+	for (i = 0; i < bd->bd_asp_nr; i++)
+		blob->bb_arena_last[i] = bd->bd_asp[i].as_arena_last;
+out:
+	return rc;
+}
+
+static uint64_t
+blob_incarnation(struct ad_blob *blob)
+{
+	return blob->bb_df->bd_incarnation;
+}
+
+static void
+blob_set_opened(struct ad_blob *blob)
+{
+	struct ad_blob_df *bd = blob->bb_df;
+
+	bd->bd_incarnation = d_timeus_secdiff(0);
+	bd->bd_back_ptr	   = (unsigned long)blob;
+
```

Comment:
(unsigned long) is 4 byte, bd_back_ptr is uint64, should we use uint64_t here?

Reply:
No, unsigned long is 8 bytes on 64-bit machine, it has the same size as pointer, so if we want to convert it to uint64_t,
 we need to do this:
 back_ptr = (uint64_t)(unsigned long)blob;
we cannot cast pointer to uint64_t directly, compiler will complain.

Reply:
you are right, i thought 'unsigned long' is different with 'unsigned long long', but they are both 8 bytes..